File: Java/src/main/java/com/gildedrose/GildedRose.java
package com.gildedrose;

class GildedRose {
    Item[] items;

    public GildedRose(Item[] items) {
        this.items = items;
    }

    public void updateItems() {
        for (Item item : items) {
            if (item.name.equals("Sulfuras, Hand of Ragnaros")) {
            } else {
                item.sellIn--;
                if (item.name.equals("Aged Brie") )
                    item.quality = Math.min(50, item.quality + (item.sellIn < 0? 2 : 1 ));
                else if (item.name.equals("Backstage passes to a TAFKAL80ETC concert"))

                    item.quality = Math.min(50, item.quality + 1);
                    if (item.sellIn < 10) {
                        if (item.quality < 50) {
                            item.quality++;
                        }
                    }
                    if (item.sellIn < 5) {
                        if (item.quality < 50) {
                            item.quality++;
                        }
                    }
                    if (item.sellIn < 0) {
                        item.quality = 0;
                    } else
                    item.quality = Math.max(0, item.quality - (item.sellIn < 0? 2 : 1 ));
            }
        }
    }
}

File: js-jest/src/gilded_rose.js
class Item {
  constructor(name, sellIn, quality) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

class Shop {
  constructor(items = []) {
    this.items = items;
  }
  updateQuality() {
    this.items.forEach((item) => {
      if (item.name === "Sulfuras, Hand of Ragnaros") {
        return;
      }
      item.sellIn = item.sellIn - 1;

      if (item.name === "Aged Brie") {
        this.calculateNewAgedBrieQuality(item);
        return;
      }

      if (item.name === "Backstage passes to a TAFKAL80ETC concert") {
        this.calculateNewBackstagePassQuality(item);
        return
      }
      if(item.name.startsWith("Conjured")){
        this.calculateNewConjuredQuality(item);
        return;
      }

      this.calculateNewNormalItemQuality(item);
    });
  }

  calculateNewConjuredQuality(item) {
    this.calculateNewNormalItemQuality(item);
    this.calculateNewNormalItemQuality(item);
  }
  calculateNewNormalItemQuality(item) {
    item.quality = Math.max(item.sellIn < 0 ? item.quality - 2 : item.quality - 1, 0);
  }

  calculateNewBackstagePassQuality(item) {
    item.quality = Math.min(item.quality + calculateOffset(), 50)
    function calculateOffset() {
      switch (true) {
        case (item.sellIn < 0):
          return (item.quality * -1);
        case (item.sellIn < 5):
          return 3;
        case (item.sellIn < 10):
          return 2;
        default:
          return 1;
      }
    }
  }

  calculateNewAgedBrieQuality(item) {
    item.quality = Math.min(item.sellIn < 0 ? item.quality + 2 : item.quality + 1, 50);
  }
}

module.exports = {
  Item,
  Shop,
};
